### ISSUE TITLE
Enhancement: [no-unsafe-*] misses `any` violations on union and object types

### ISSUE THREAD

**Issue by ronami** [typescript-eslint/typescript-eslint#10158]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.3.2&fileType=.tsx&code=PTAEDsHsFoFdwM4EMBmBTaSBOBzWBbNcAFwC4AocgEzQGMAbbNUWyRY0JU0Ab1BUiRuScAE9QAXwDclFPFrEAlm36CAFAA9ufAUNAJiWReBySAlNtV6DRk5N7lQoLGmKws4UBpkTKIUACqiKjM2HiEJKCQKKDEogAOzAAGOoLCYlKSSZwICIo44GhUsZCcoPHYSITEaFhRMXGJoClW3DbGOJkSSQB05LpqSGYy1HSMLixsBqAARumiANoAuqAAPhAEM7UjcuAKyp4z2Jpthh3Laxv4W1gW%2BmcmF%2Bvgm7UOTi5uHl4%2BfmBByHQnFwBCIHGisQSyRe1ze6xEiyW2SQuXyhWKxFKSHKlWqbwhjWhrzq63ajyRfSOWDUM2Gf30inwinG9HEAHdFMQABbNKBwYLoaCfdzgJKUGgMJiTdgsSy6eaSEZoDTxSBYDi7fYqLloej0SBqO6pawPUwSd6gfwAkLOVwi%2BqQppJBHZFBYSD4fjyJQqDnc21fTyE5pknC9RwB%2B1IX6jSUTVgyqjzJ5XG5KlVqjXeg6gNlq%2BhUQ2nWw4FMwm4Wq0C5jC74EqHNBHLV3uz2an2eP082tBhtJUPNvofO3fGY%2BIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolp9kySgHMmAW0Qt0URNGgdokADThseIqWQVqdBszaduvQcOiJ8saEwWQlKtVoC%2BIN0A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
// no-unsafe-argument:

declare const a: { foo: any };

function foo(x: { foo: string }): { foo: string } {
  return x;
}

// Unsafe argument of type `{ foo: any; }` assigned to a parameter of type `{ foo: string; }`.
foo(a);

declare const b: any[] | number;

function bar(x: string[] | number): string[] | number {
  return x;
}

// Unsafe argument of type `number | any[]` assigned to a parameter of type `number | string[]`.
bar(b);

// similarly with `no-unsafe-return`

declare const c: { foo: any };

export function hello(): { foo: string } {
  // Unsafe return of type `{ foo: any }` from function with return type `{ foo: string }`.
  return a;
}

declare const d: any[] | number;

export function world(): string[] | number {
  // Unsafe return of type `any[]` from function with return type `number | string[]`.
  return b;
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unsafe-assignment": "error",
    "@typescript-eslint/no-unsafe-return": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected the rule to fail on these, similarly to how it fails on similar boxed values like `Promise<any>`, or `any[]`, along with a combination of boxed values.

### Actual Result

The code example is reporting no lint errors.

### Additional Info

The `no-unsafe-*` rule family misses some edge case checks, mostly around boxed values (like arrays, objects, promises, etc.). While you get a warning for most `any` values wrapped in a box, I found some the rule misses around union and object types.

I made a simplified implementation to test-run on typescript-eslint's own repo and [found one issue here](https://github.com/typescript-eslint/typescript-eslint/blob/e2d4347f2e70228fb143008bceb139d8898ec4dc/packages/website/src/components/config/ConfigTypeScript.tsx#L53) (matching `string[] | undefined` with `any[]`).

**Comment by ronami** [comment]
Adding that functions are also being missed:

```ts
declare const onSomething: (callback: () => number) => void;
declare const callback: () => any;

// no lint errors
onSomething(callback);
```

**Comment by bradzacher** [comment]
The rules currently do not recurse into cases like arguments, return types, or sub-union members - for complexity reasons on the initial implementation and generally these cases are caught at the declaration site.

Eg the return type issue should be handled at the function declaration via `no-unsafe-return`. The argument issue (or technically any of these issues) is caught by `no-explicit-any`.

Usually the only time you'd be dealing with such cases like these is if an external library type has given them to you.

Do you have a real-world example of this that a combination of all of the current `no-unsafe` rules plus `no-explicit-any` wouldn't handle?

**Comment by ronami** [comment]
> The rules currently do not recurse into cases like arguments, return types, or sub-union members - for complexity reasons on the initial implementation and generally these cases are caught at the declaration site.
> 
> Eg the return type issue should be handled at the function declaration via `no-unsafe-return`. The argument issue (or technically any of these issues) is caught by `no-explicit-any`.
> 
> Usually the only time you'd be dealing with such cases like these is if an external library type has given them to you.
> 
> Do you have a real-world example of this that a combination of all of the current `no-unsafe` rules plus `no-explicit-any` wouldn't handle?

Yes, @bradzacher, I agree that this error will usually pop up if you use a badly typed external library. I didn't encounter a real-world example; this mostly started as a discussion at https://github.com/typescript-eslint/typescript-eslint/pull/10051#discussion_r1799961968.

Out of curiosity, I did write a rough implementation of this and ran it on `typescript-eslint`'s codebase (not including checking function types, just checking unions recursively). I found one [valid issue here](https://github.com/typescript-eslint/typescript-eslint/blob/e2d4347f2e70228fb143008bceb139d8898ec4dc/packages/website/src/components/config/ConfigTypeScript.tsx#L53). A similar issue did pop two more times: [1](https://github.com/typescript-eslint/typescript-eslint/blob/bce4560cb04177c9f0821709e9a69a55f48a5707/packages/website/src/components/lib/jsonSchema.ts#L191), [2](https://github.com/typescript-eslint/typescript-eslint/blob/bce4560cb04177c9f0821709e9a69a55f48a5707/packages/website/src/components/lib/jsonSchema.ts#L199). Here's [a playground link with a smaller repro](https://typescript-eslint.io/play#ts=5.4.5&fileType=.tsx&code=JYOwLgpgTgZghgYwgAgPIAczAPYjgGwEFkBvAKGUuRmAnwBMBnAfgC5lGwpQBzAbQC6AbjIBfMmVCRYiFADlsYDFlwFi5SjTpN2nbiH7CxE%2BhAT44UFAlydkcdhuTZMOPEUbtlbtYJGUQRW9VD3YFJVcQwj8xETJTc0trWzBkACN2OBAATxiyAHp85HQ4RkYIRmQACkYAC2wAVwZqOGB8AEoyOAA6FxV3QkZu9Aa6qqctBk90gBoxdrjClrbGLu7AiP61IZGxidop9jS50QWyIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tieQEMAZolp9oAc1gBbRC3RRE0aB2iQANOGx4ipZBWp0GzNp269Bw6InyxoTOZAVKV67FAIlyVGvWSMW7Lh5%2BIVppSQAjBREyMhRUDAdFZTUNHHcdPW9DfxMg8xFkZEoxJmlZBMdkjQBfEGqgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false).

I'll add that on my (potentially faulty) implementation, the time it takes to lint `typescript-eslint`'s entire project did increase from ~51 to ~64 seconds on my local machine.

I don't know if this is worth the extra complexity and time the rule will take to execute.

**Comment by bradzacher** [comment]
The extra time would be because all type information is computed lazily - the more you access, the more that gets computed.

So by going deeper into the inspections of things it doesn't surprise me that the rule gets slower!

**Comment by JoshuaKGoldberg** [comment]
Tempting, but I'm not in favor. I get the appeal of more deeply checking for `any`s, but if those nested properties actually get used, then I think they would get reported most of the time. The `element.type.keys()` examples are nice motivators but IME not super commonplace. This'd quite a lot of work + a bit of performance bloat for IMO not very much gain.

**Comment by ronami** [comment]
> Tempting, but I'm not in favor. I get the appeal of more deeply checking for `any`s, but if those nested properties actually get used, then I think they would get reported most of the time. The `element.type.keys()` examples are nice motivators but IME not super commonplace. This'd quite a lot of work + a bit of performance bloat for IMO not very much gain.

I like having rules be as strict as they can be, but I agree that fringe examples aside, it's not worth the extra complexity and the performance hit.

The issue has been open for just over two and a half months with no community reactions.

In my opinion, this can be closed 👍
